src/auth/src/credentials/user_credential.rs

```diff
@@ -14,9 +14,70 @@
 
 use crate::credentials::traits::dynamic::Credential;
 use crate::credentials::Result;
-use crate::errors::CredentialError;
+use crate::errors::{is_retryable, CredentialError};
 use crate::token::{Token, TokenProvider};
-use http::header::{HeaderName, HeaderValue, AUTHORIZATION};
+use http::header::{HeaderName, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
+use reqwest::{Client, Method};
+use std::time::Duration;
+use time::OffsetDateTime;
+
+#[allow(dead_code)] // TODO(#442) - implementation in progress
+struct UserTokenProvider {
+    client_id: String,
+    client_secret: String,
+    refresh_token: String,
+    endpoint: String,
+}
+
+#[async_trait::async_trait]
+impl TokenProvider for UserTokenProvider {
+    async fn get_token(&mut self) -> Result<Token> {
+        let client = Client::new();
+
+        // Make the request
+        let req = Oauth2RefreshRequest {
+            grant_type: RefreshGrantType::RefreshToken,
+            client_id: self.client_id.clone(),
+            client_secret: self.client_secret.clone(),
+            refresh_token: self.refresh_token.clone(),
+        };
+        let header = HeaderValue::from_static("application/json");
+        let builder = client
+            .request(Method::POST, self.endpoint.as_str())
+            .header(CONTENT_TYPE, header)
+            .json(&req);
+        let resp = builder
+            .send()
+            .await
+            .map_err(|e| CredentialError::new(false, e.into()))?;
```

Comment:
Maybe for a future PR:

Depending on the nature of the error this may be retryable.  DNS resolution and connections refused are typically retryable.  I cannot think of any I/O error that is not retryable for these purposes.  Can you?